Added container support

XedinUnknown opened the pull request:
Now working with any readable/writable containers.

Aims to incorporate solution for #6.

XedinUnknown committed:
Corrected return type of `_getDataStore()`

XedinUnknown committed:
Changed expectations of `GetDataCapableTrait` and `HasDataCapableTrait`

XedinUnknown committed:
`GetDataCapableTrait` and `HasDataCapableTrait` now abstract container operations

XedinUnknown committed:
Expanded range of returned datastore types

XedinUnknown committed:
`_setData()` now expected to abstract container operation

XedinUnknown committed:
Added suggested package

XedinUnknown committed:
`_setData()` now expected to throw 3 types of exception

XedinUnknown committed:
`_setData()` now abstracts container access and throws 3 exceptions

XedinUnknown committed:
Now mocking the i18n method

XedinUnknown committed:
`_setData()` now internationalizes exception messages

XedinUnknown committed:
Corrected throws doc

XedinUnknown committed:
`_setMany()` now expected to abstract container access and throw 3 exceptions

XedinUnknown committed:
`_setMany()` now abstracts container operations and throws 3 exceptions

XedinUnknown committed:
`_unsetData()` now expected to abstract container operations and throw 4 exceptions

XedinUnknown committed:
`_unsetData()` now abstracts container operations and throws 4 exceptions

XedinUnknown committed:
`_unsetMany()` is now expected to abstract container operation and throw 4 types of exceptions

XedinUnknown committed:
`_unsetMany()` now abstracts container operations and throws 4 types of exception

XedinUnknown committed:
`_getData()` now expected to throw `InvalidArgumentException` for invalid key

Also tested other paths

XedinUnknown committed:
`_getData()` now throws `InvalidArgumentException` for invalid key

XedinUnknown committed:
Now mocking i18n method

XedinUnknown committed:
Changed expectations of `DataStoreAwareTrait`

- It must be named `DataStoreAwareContainer.
- Must store a readable container.

XedinUnknown committed:
Changed `DataStoreAwareTrait`

- Renamed.
- Now stores readable container.

mecha commented on src/DataStoreAwareContainerTrait.php:
```diff
@@ -0,0 +1,66 @@
+     * @since [*next-version*]
+     *
+     * @param array|ArrayAccess|stdClass|ContainerInterface|null $dataStore A container.
+     */
```
This method uses `_normalizeContainer()`, which can throw `InvalidArgumentException`. Shouldn't this method also declare that it `throws` an `InvalidArgumentException`?

XedinUnknown replied:
Yes.

mecha committed:
Fixed PHPUnit config for PHPS

mecha committed:
Ignoring more irrelevant PHPS files

mecha committed:
Added missing `throws` doc tag

The `_normalizeContainer()` abstract method can throw `InvalidArgumentException`.
Therefore, since `_setDataStore()` does not catch it, it too should document it.
